### PR TITLE
Automatic template insertion in daily notes

### LINKED ISSUE #163: Use a specific template with daily or new note?
### 🚀 The feature, motivation and pitch

Other than using :ObsidianTemplate and inserting a template, is there a way to create a new daily or note from a template?

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by MunsMan**
This pull request adds support for automatic template insertion in daily notes.

To enable the feature, just add the full template name to the daily note config:

```lua

{
    daily_note = {
         template = "daily.md"
    }
}

```

Furthermore, I added support for custom template variables.

It is done by the `util.substitute_template_variables` with tasks the currently available variables and a map of new variables.
There the variables are a name, function pair. The executed function should return the inserted value.
An Example is in the README. 

Currently, custom variables aren't added to the normal templates yet, but that would be just a three line change.
Please give me your thought on this feature.

This should close #163

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -323,7 +327,8 @@ For example, with the following configuration
   templates = {
       subdir = "my-templates-folder",
       date_format = "%Y-%m-%d-%a",
-      time_format = "%H:%M"
+      time_format = "%H:%M",
+      custom = {}
```
Comment (epwalsh) [inline README.md:331]: How about calling this `substitutions` (or something like that)?
```suggestion
	substitutions = {}
```

**Review by epwalsh — COMMENTED** [review]
Hey @MunsMan, this is great. 

> Currently, custom variables aren't added to the normal templates yet, but that would be just a three line change.

I think it makes sense to add that in this PR, do you mind?

File: README.md [hunk 2]
```diff
@@ -214,6 +216,8 @@ This is a complete list of all of the options that can be passed to `require("ob
     subdir = "templates",
     date_format = "%Y-%m-%d-%a",
     time_format = "%H:%M",
+	-- A map for custom variables, the key should be the variable and the value a function
+	custom = {}
```
[the remark this answers is not in the document]
Comment (MunsMan) [reply README.md:220]: I thought that are the default options. I would list it there with the empty list, just to make the exists of the option clear. Likewise, I will update it to `substitutions`.
Please correct me if I'm wrong. I thought the default config is from line 135 to 258 and on line 324 to 333 are an example. I will add a proper `substitutions` example.

**Comment by MunsMan** [comment]
I just added the substitution support for the default template as well. Furthermore, I renamed it to `substitutions` and cleaned one spot in the README.

If you have anything else, just led me know.

File: lua/obsidian/init.lua [hunk 3]
```diff
@@ -385,6 +385,10 @@ client._daily = function(self, datetime)
   -- Create Note object and save if it doesn't already exist.
   local note = obsidian.note.new(id, { alias }, { "daily-notes" }, path)
   if not note:exists() then
+    if self.opts.daily_notes.template then
+      obsidian.util.clone_template(self.opts.daily_notes.template, tostring(path), self, note:display_name())
+      note = obsidian.note.new(id, { alias }, { "daily-notes" }, path)
```
Comment (epwalsh) [inline lua/obsidian/init.lua:390]: This is the same as above on line 386? Looks like we don't need this.
```suggestion
```

File: README.md [hunk 4]
```diff
@@ -344,7 +348,22 @@ Date created: 2023-03-01-Wed
 
 above the cursor position.
 
-## Known Issues 
+To further customize to your personal preferences, you can supply a custom map with your on template variables.
```
Comment (epwalsh) [inline README.md:351]: How about:
```suggestion
You can also define custom template substitutions with the configuration field `templates.substitutions`. For example, to automatically substitute the template variable `{{yesterday}}` when inserting a template, you could add this to your config:
```

File: README.md [hunk 5]
```diff
@@ -344,7 +348,22 @@ Date created: 2023-03-01-Wed
...
+{
+    -- other fields ...
+   templates = {
+        substitutions = {
+            yesterday = function()
+                return os.date("%Y-%m-%d", os.time() - 86400)
+            end
+        }
+    }
+}
```
Comment (epwalsh) [inline README.md:363]: Let's keep the indent consistent with 2 spaces:
```suggestion
-- other fields ...
templates = {
  substitutions = {
    yesterday = function()
      return os.date("%Y-%m-%d", os.time() - 86400)
    end
  }
}
```

File: README.md [hunk 6]
```diff
@@ -214,6 +216,8 @@ This is a complete list of all of the options that can be passed to `require("ob
     subdir = "templates",
     date_format = "%Y-%m-%d-%a",
     time_format = "%H:%M",
+	-- A map for custom variables, the key should be the variable and the value a function
+	substitutions = {}
```
Comment (epwalsh) [inline README.md:220]:
```suggestion
    -- A map for custom variables, the key should be the variable and the value a function
    substitutions = {}
```

File: README.md [hunk 7]
```diff
@@ -150,6 +150,8 @@ This is a complete list of all of the options that can be passed to `require("ob
     date_format = "%Y-%m-%d",
     -- Optional, if you want to change the date format of the default alias of daily notes.
     alias_format = "%B %-d, %Y"
+    -- Optional, if you want to automaticly insert a template from your template directory like 'daily.md'
```
Comment (epwalsh) [inline README.md:153]:
```suggestion
    -- Optional, if you want to automatically insert a template from your template directory like 'daily.md'
```

File: CHANGELOG.md [hunk 8]
```diff
@@ -10,6 +10,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 
 - Added an extra option for daily notes to allow changing the default title from "%B %-d, %Y" to other strings.
+- Added an option for daily notes to automatically load a template.
+- Adding an option for custom template variables.
+  - Currently, only supported in the daily note template insertion.
```
Comment (epwalsh) [inline CHANGELOG.md:15]: How about:
```suggestion
- Added a configuration option `daily_notes.template` for automatically using a specific template when creating a new daily note.
- Adding a configuration option `templates.substitutions` for defining custom template substitutions.
```

**Review by epwalsh — APPROVED** [review]
Thanks @MunsMan!

**Pull request merged**